### PR TITLE
[DeepSeek V4]Add Deepseek v4 indexer

### PR DESCRIPTION
**Pull request by inucool12**
# DeepSeek-V4 Lightning Indexer: TPU StreamIndex Top-K Kernel & Indexer

## Overview
  This design implements the **streamindex_topk** custom kernel and its PyTorch/vLLM integration module **VllmDeepseekV4Indexer** for the **DeepSeek-V4 Lightning Indexer**.
  
  The integration operates across three main components:
  1. **PyTorch Indexer Layer (`deepseek_v4_indexer.py`)**: A vLLM-compatible indexer module that intercepts forward passes, structures and buckets metadata (batch, token, block dimensions) to avoid compilation overhead, and performs JAX interop using TorchAX.
  2. **JAX Pallas Kernel (`streamindex_topk.py`)**: A hardware-optimized custom kernel on TPUs mapping sequences directly to hardware program grids (`pl.program_id`). It performs running Top-K computations fully on-chip in Vector Memory scratchpads.
  3. **Mocks & Tests (`test_streamindex_topk.py`)**: Validates shape constraints and verifies batched numerical correctness against a naive NumPy implementation.

### CODE REVIEW

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+      compressed_kv_score: torch.Tensor,
+      indexer_weights: torch.Tensor,
+      positions: torch.Tensor,
+      rotary_emb: nn.Module,
+      slot_mapping: Optional[torch.Tensor] = None,
+  ) -> torch.Tensor:
+
+    actual_num_tokens = hidden_states.shape[0]
+
+    if actual_num_tokens == 0:
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: nit: we could remove " if actual_num_tokens == 0:..."
forward () won't be called if there is no tokens for this step.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+# limitations under the License.
+"""TPU-compatible DeepSeek-V4 Lightning Indexer."""
+
+from typing import Optional, Tuple
+
+import torch
+import torch.nn as nn
+from torchax.interop import jax_view
+from torchax.interop import torch_view
+from tpu_inference.layers.vllm.custom_ops.streamindex_topk import streamindex_topk
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: the import should from kernel directory(tpu_inference/kernels/experimental/deepseek_v4)?

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    # TODO(hwanginho): Support DSv4 FP8 index cache format.
+    # Once vLLM's FP8 cache is enabled, extract the scale tensor here if it
+    # is a separate tensor, or pack it with cache_kv. Need to align with
+    # alynie@ on https://github.com/vllm-project/tpu-inference/pull/2858
+    # (The exact attribute name depends on vLLM's FP8 implementation, e.g., `kv_scale`)
+    # kv_scales_tensor = self.k_cache.kv_scale
+    # assert kv_scales_tensor.is_contiguous(), "Scales must be contiguous"
+    # kv_scales_jax = jax_view(kv_scales_tensor)
+
+    c_s = getattr(self, "c_s", 8192)
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: these are stale, we no longer need them?

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 1 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Removed

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 3 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Removed

**Comment by inucool12** [comment]
@gxd3 , I have cleaned up the code by removing all of the manual padding and alignment logic, since these operations are already pre-applied upstream before the indexer hook is called.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  ) -> torch.Tensor:
+
+    actual_num_tokens = hidden_states.shape[0]
+
+    q, _ = self.wq_b(query)
+    q = q.view(-1, self.n_head, self.head_dim)
+
+    q_quant, q_scales = fused_indexer_q_rope_quant(q, positions, rotary_emb)
+
+    # Do not fold q_scales into weights. Keep weights static to save HBM
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: why we don't fold `q_scales` into `weights`?
vLLM does that: https://screenshot.googleplex.com/56K2NH8dUeMmNsZ

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    # 3. DIRECT JAX KERNEL CALL
+    # ---------------------------------------------------------
+    bt_slice = attn_metadata.block_table  # pytype: disable=attribute-error
+    sl_slice = attn_metadata.seq_lens  # pytype: disable=attribute-error
+    csl_slice = attn_metadata.cu_seq_lens  # pytype: disable=attribute-error
+
+    topk_indices = streamindex_topk(
+        query_projection=q_quant[:actual_num_tokens],
+        kv_cache=kv_cache_tensor,
+        block_table=bt_slice,
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: page_indices=...?

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        seq_lens=sl_slice,
+        cu_q_lens=csl_slice,
+        indexer_weights=weights,
+        k=self.topk_tokens,
+        compression_ratio=self.compress_ratio,
+    )
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: num_kv_pages_per_block 
num_queries_per_block
need to be set,
we could set them to 1 for now and leave a TODO to tune them later on.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+      rotary_emb: The vLLM RoPE CustomOp module (Intercepted by TPU rope.py)
+
+  Returns:
+      q_quant: Int8 quantized, RoPE-applied queries
+      q_scales: Quantization scales
+  """
+
+  # We clone `q` so we don't mutate the original query.
+  # XLA will optimize the clone away.
+  q_rope, _ = rotary_emb(positions, q.clone())
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: why "clone"?

Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: this is q (including nope dimension & rope dimension)?

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+  # Find absolute max along the head dimension
+  amax = torch.amax(torch.abs(q_rope), dim=-1, keepdim=True)
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: can use "quantize_tensor" from
`from tpu_inference.layers.common.quantization import quantize_tensor`

I believe vLLM's implemention round up the scale to the next power of two:
https://screenshot.googleplex.com/3LSsLTUQ7MZtwzd

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 4 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Done. I have updated the code to fold `q_scales` directly into the indexer weights before calling the JAX kernel.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 5 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Typo. Fixed

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    q, _ = self.wq_b(query)
+    q = q.view(-1, self.n_head, self.head_dim)
+
+    q_quant, q_scales = fused_indexer_q_rope_quant(q, positions, rotary_emb)
+
+    weights = (
+        indexer_weights.to(q.dtype) * self.softmax_scale * (self.head_dim**-0.5)
+    )
+    # Fold the query quantization scales into the weights
+    weights = weights * q_scales.unsqueeze(-1)
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: nit: merge into line 102?

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+      positions: Token positions of shape [num_tokens]
+      rotary_emb: The vLLM RoPE CustomOp module (Intercepted by TPU rope.py)
+
+  Returns:
+      q_quant: Int8 quantized, RoPE-applied queries
+      q_scales: Quantization scales
+  """
+
+  # Apply the custom rotary embedding op directly in-place on the query tensor.
+  q_rope, _ = rotary_emb(positions, q)
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: nit: this is q (including both nope and rope dimension?)

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 6 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Agreed. I set both to 1 for now and add a TODO to tune these block size parameters later.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 7 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Acually, no need to clone. Updated.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 8 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Updated

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 7 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: Yes, this is the entire query q of the indexer.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 10 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: > nit: this is q (including both nope and rope dimension?)
Yes, this is the entire query q of the indexer .

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 9 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: > nit: merge into line 102?

Good catch. Done.

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 10 above)
Comment (gxd3) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: maybe update the naming as well?

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 8 above)
Comment (gxd3) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: is the code not synced?
didn't see `quantize_tensor` still

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 10 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: > maybe update the naming as well?

Done

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 8 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: > is the code not synced? didn't see `quantize_tensor` still
Ah, I see. I have updated the code to call `quantize_tensor` from `tpu_inference.layers.common.quantization` using TorchAX (`jax_view`/`torch_view`).

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+      q_quant: Int8 quantized, RoPE-applied queries
+      q_scales: Quantization scales
+  """
+
+  # Apply the custom rotary embedding op directly in-place on the query tensor.
+  q, _ = rotary_emb(positions, q)
+
+  # Bridge to JAX, call JAX quantize_tensor, and bridge back to PyTorch
+  q_jax = jax_view(q)
+  q_quant_jax, q_scales_jax = quantization.quantize_tensor(q_jax, jnp.int8)
```
Comment (gxd3) [inline tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: this should be fp8?

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 8 above)
Comment (gxd3) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: for the q scale:
I believe vLLM's implemention round up the scale to the next power of two:
https://screenshot.googleplex.com/3LSsLTUQ7MZtwzd

File: tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py (hunk 11 above)
Comment (inucool12) [reply tpu_inference/layers/vllm/custom_ops/experimental/deepseek_v4/deepseek_v4_indexer.py]: You are right. Fixed

**Review by gxd3 — APPROVED** [review]

**Comment by inucool12** [comment]
@a1yssan13 , I updated deepseek_v4_indexer.py by integrating your compressor implementation. 
Please review this when you have a moment!

File: tests/kernels/deepseek_v4/test_streamindex_topk.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+  print(f"  - GQA: {H_I} Query Heads -> {H_KV} KV Heads")
+  print(f"  - Compression Ratio: {comp_ratio}")
+
+  print("\nInputs Generated:")
+  print(f"  - Query Tensor: {q.shape}")
+  print(f"  - KV Cache:     {kv.shape}")
+  print(f"  - Block Table:  {block_table.tolist()}")
+
+  # =====================================================================
+  # 2. NAIVE COMPUTATION (The Ground Truth)
```
Comment (a1yssan13) [inline tests/kernels/deepseek_v4/test_streamindex_topk.py]: nit: it might be cleaner to to have a seperate method streamindex_topk_ref as a reference implementation to compare against?

**Review by a1yssan13 — APPROVED** [review]
thanks!
